ADT7420 Linux Support and Linux I2C API Fix

arthur428 opened the pull request:
## Pull Request Description
Relevant Issue: [2518](https://github.com/analogdevicesinc/no-OS/issues/2518)
- Added ADT7420 Linux platform support
- Re-implementation of Linux I2C read and write APIs for `stop_bit` handling

Tested with the adt7420-pmdz dummy example and an ADT7420 + RPI CM5 (RPI OS) setup.

## PR Type
- [x] Bug fix (change that fixes an issue)
- [x] New feature (change that adds new functionality)
- [ ] Breaking change (has dependencies in other repos or will cause CI to fail)

## PR Checklist
- [x] I have followed the [Coding style guidelines](http://analogdevicesinc.github.io/no-OS/drivers_guide.html#coding-style)
- [x] I have complied with the [Submission Checklist](http://analogdevicesinc.github.io/no-OS/contributing.html#submission-checklist)
- [x] I have performed a self-review of the changes
- [x] I have commented my code, at least hard-to-understand parts
- [ ] I have build all projects affected by the changes in this PR
- [x] I have tested in hardware affected projects, at the relevant boards
- [x] I have signed off all commits from this PR
- [x] I have updated the documentation (wiki pages, ReadMe etc), if applies

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/analogdevicesinc/no-OS?pullRequest=2593) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/analogdevicesinc/no-OS?pullRequest=2593) before we can accept your contribution.<br/>**0** out of **2** committers have signed the CLA.<br/><br/>:x: jmacagba<br/>:x: arthur428<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/analogdevicesinc/no-OS?pullRequest=2593) it.</sub>

amiclaus commented on drivers/platform/linux/linux_i2c.c:
```diff
@@ -50,8 +50,89 @@
+ * @param read - 0 to write, otherwise read. 
+ * @return 0 in case of success.
+ */
+int32_t linux_i2c_add_msg(struct no_os_i2c_desc *desc,
```
`int` as return type for functions.

amiclaus commented on drivers/platform/linux/linux_i2c.c:
```diff
@@ -50,8 +50,89 @@
+	msg.len = bytes_number;
+	msg.buf = data;
+
+	if (read) {
```
drop brackets if only one statement is guarded. check all occurences.

amiclaus commented on drivers/platform/linux/linux_i2c.c:
```diff
@@ -50,8 +50,89 @@
+	packets.nmsgs = linux_desc->len_messages;
+
+	ret = ioctl(linux_desc->fd, I2C_RDWR, &packets);
+	if (ret < 0) {
```
`if (ret)`

amiclaus commented on drivers/platform/linux/linux_i2c.c:
```diff
@@ -231,7 +231,7 @@ int32_t linux_i2c_write(struct no_os_i2c_desc *desc,
 		printf("%s: Can't allocate memory\n\r", __func__);
 		return -1;
 	}
-	
```
redundant change.

amiclaus reviewed (commented):
most of the commits do not follow the guidelines: https://analogdevicesinc.github.io/no-OS/contributing.html

arthur428 commented:
Hi, I just discovered a breaking bug on my latest commit.

After changing all the `if (ret < 0)` instances to `if (ret)`, I was not able to run the ADT7420-pmdz sample project. It turns out the problem was the `if` statement inside `linux_i2c_send_msg` after the `ioctl()` call.

It looks like `ioctl(..., I2C_RDWR, ...)` returns how many messages were successfully sent. Even in a successful execution, `linux_i2c_send_msg` will return `-1`.

I'll upload a new commit and also include the fixes suggested by the testing pipeline.

arthur428 commented:
Test Logs for ADT7420 Sample Project on RPI CM5 + ADT7420 (I2C)
Commit: b9f754e
Duration: ~2.5h
Logs: [may_20_2025_b9f754e.log](https://github.com/user-attachments/files/20325785/may_20_2025_b9f754e.log)

arthur428 commented on drivers/platform/linux/linux_i2c.c:
```diff
@@ -95,6 +95,7 @@ int linux_i2c_add_msg(struct no_os_i2c_desc *desc,
 		linux_desc->messages = ptr;
 	} else {
 		linux_desc->messages = no_os_malloc(sizeof(struct i2c_msg));
+		linux_desc->len_messages = 0;
```
Tested with a ROS2 driver running on Ubuntu 22, ROS2 Humble. Without this line, `len_messages` does not get initialized to 0, causing a segmentation fault.

arthur428 commented:
Hi, just following up on this PR.

Is there a specific timeline for merging? Also is there anything else I need to do for the PR to be accepted?

Thank you.

buha commented:
ok so 1 general comment, use `git commit -s` to add that Signed-off-by: line at the end of each commit body

buha commented:
another general comment, you have 9 commits but most of them are fixes and review related changes, just squash them into 1 or 2 commits as you see fit that should contain everything

buha commented on projects/adt7420-pmdz/src/examples/dummy/dummy_example.c:
```diff
@@ -62,13 +62,15 @@ int example_main()
 	float temp_now;
 	int ret;
 
+#ifndef LINUX_PLATFORM
```
this commit title is `Added linux platform in project/adt7420-pmdz (untested)`, is this true ? this has to be tested

arthur428 commented:
Hi,

As advised, I've squashed the commits and separated the changes for the ADT7420 project and the Linux I2C driver. I've also added sign-off and co-author lines at the end of the commit messages. Let me know if there are still TODOs on my part.

Thank you.

buha commented:
/AzurePipelines run

buha commented:
@arthur428 i will merge it as-is, keep that in mind for the future

arthur428 commented:
Hi buha, I will keep that in mind. Thank you.
